pkg/federate/kubefed/federator.go

```diff
@@ -1,17 +1,59 @@
 package kubefed
 
 import (
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
 	"github.com/submariner-io/admiral/pkg/federate"
+	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	kubeFedNamespace string
 )
 
+func init() {
+	flag.StringVar(&kubeFedNamespace, "kubefed-namespace", "kube-federation-system", "The namespace in which the KubeFed control plane is deployed.")
+}
+
+type clusterWatcher struct {
+	eventQueue           chan func()
+	enqueueTimerInterval time.Duration
+	stopChan             <-chan struct{}
+	handler              federate.ClusterEventHandler
+	handlerString        string
+}
+
 type federator struct {
-	kubeConfig *rest.Config
+	sync.Mutex
+	clusterMap             map[string]*rest.Config
+	kubeFedClient          kubernetes.Interface
+	kubeFedClusterInformer cache.SharedIndexInformer
+	startInformerOnce      sync.Once
+	clusterWatchers        []*clusterWatcher
+	stopChan               <-chan struct{}
 }
 
-// New creates a new kubefed Federator instance.
-func New(kubeConfig *rest.Config) (federate.Federator, error) {
+// New creates a new Federator instance.
```

Comment:
can we add another "New" method with no "kubeFedConfig" parameter, which would call this one with:

client-go.rest.config.InClusterConfig as rest.Config?

I suspect that in such case, we can save the hurdle of passing config/credentials if we co-located it federation resources. Although we should always allow passing the config options for federation.

Reply:
The typical usage pattern I've seen is:

```
func init() {
	flag.StringVar(&kubeConfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
}
...
kubeConfig, err := clientcmd.BuildConfigFromFlags(masterURL, kubeConfig)
```
So `BuildConfigFromFlags` uses _kubeConfig_ if non-empty, otherwise _masterURL_ if non-empty otherwise it uses _InClusterConfig_. So a `New` method with no  _kubeConfig_ parameter probably wouldn't be useful. We could pass in _kubeConfig_ and _masterURL_  or nothing and have `New` extract the command args but I figured it's better to have the app do that part as it will likely need to obtain a _kubeConfig_ for other purposes.

Reply:
You're right. BuildConfigFromFlags already handles the incluster config if those flags aren't passed.

Reply:
Can we add the example you pasted here to docs?, or provide an example file somewhere?

I was struggling over this when trying to do it on coastguard.

Reply:
yeah even though this is "basic" k8s stuff, it's not easy nor intuitive. I used subM code as a guide. I did this in the orchestrator code which I'll be publishing soon.

Reply:
yeah, import side effects create some counter intuitive patterns... not Tom's fault though